Script interaction in python tests (prototype)

AnOctopus opened the pull request:
A mostly working but hacky implementation of forward message parsing, element query and interaction apis, and test script runner apis. One cypress test file replaced. 

Getting the value of a widget that was interacted with does not work correctly, we will fix it in the full implementation.

Probably not worth actually merging in its current state, creating this as a PR primarily to make it trivial to view as a diff and to discuss it.

AnOctopus committed:
Initial implementation of element tree parsing

With copy-pasted test script runner for now, until I can clean it up

AnOctopus committed:
Text and Radio elements

AnOctopus committed:
Pack scriptrunner boilerplate into `run` method

AnOctopus committed:
Smarter script waiting, refs to root in tree

AnOctopus committed:
Everything has ref to root

AnOctopus committed:
Root class

AnOctopus committed:
Concise, fluent api

AnOctopus committed:
Test imports cleanup

AnOctopus committed:
Not actually using AsyncTestCase capabilities

AnOctopus committed:
Remove unused annotations import

AnOctopus committed:
Hacky inline script support

AnOctopus committed:
Remove use of undefined variable

AnOctopus committed:
Rename function for running script by filename

AnOctopus committed:
Cleanup unused function and comment pluralization

AnOctopus committed:
Shorter var name in tests

AnOctopus committed:
Remove now-redundant cypress test

AnOctopus committed:
More direct script stopped detection logic

AnOctopus committed:
More tests, rename TestScriptRunner, fill out radio class api

AnOctopus committed:
Rename interactive scripts test file

AnOctopus committed:
Test local variable rename

AnOctopus committed:
Delete dead code

AnOctopus committed:
Add note about modified widget values

AnOctopus committed:
Test tree iteration order

AnOctopus committed:
`Literal` is in typing_extensions in 3.7

AnOctopus committed:
Merge branch 'develop' into prototype/unit-test-interaction

AnOctopus committed:
Use future annotations features

AnOctopus committed:
Indexing into blocks

AnOctopus committed:
Basic support for lookup by widget's key

AnOctopus committed:
Test missing widget key

AnOctopus committed:
Allow inline scripts to be indented

AnOctopus committed:
Look up current widget values in session state

AnOctopus committed:
Remove text workaround for viewing widget state

AnOctopus committed:
Merge branch 'develop' into prototype/unit-test-interaction

AnOctopus committed:
Merge branch 'develop' into prototype/unit-test-interaction

AnOctopus commented on lib/tests/interactive_scripts.py:
```diff
@@ -0,0 +1,426 @@
+
+
+@dataclass(init=False)
+class Radio(Element):
```
Unclear if we should use properties for the protobuf-delegating accessors, or if we should extract them into fields in the constructor. Some perf differences but probably not worth worrying about, so mostly a question of clarity

AnOctopus committed:
`script_from_string` as test class method

And move that test class into the `interactive_scripts` module to be subclassed

AnOctopus committed:
Fix leftover test code not using getitem

AnOctopus committed:
Remove dead patching line

AnOctopus committed:
Move other script loading function into class

AnOctopus committed:
Convert properties to fields

Much less boilerplate

AnOctopus committed:
Index based on current Radio value

Loading a stale index caused the wrong widget state to be returned

AnOctopus committed:
Value is the source of truth for Radio, not index

AnOctopus committed:
Add test for querying within a block

AnOctopus committed:
Docstrings for LocalScriptRunner

AnOctopus committed:
Fix Element value implementation, add documentation

AnOctopus committed:
Move Text static fields to declaration

AnOctopus committed:
Move Root type into declaration

AnOctopus committed:
Rename Root->ElementTree

AnOctopus committed:
Be explicit again about possible lack of fields

AnOctopus committed:
Ensure each script run has an independent copy of session state with correct values

AnOctopus committed:
Make Radio generic

AnOctopus committed:
Error if given a file that doesn't exist

AnOctopus committed:
Rename interaction files, make session_state always present

To avoid spurious type errors when accessing session_state in tests.

vdonato commented on lib/tests/interactive_scripts.py:
```diff
@@ -0,0 +1,447 @@
+# trivial methods. It may have significantly reduced scope, or be removed
+# entirely, once all elements have been implemented.
+@dataclass(init=False)
+class Element:
```
Do you think it might be reasonable to have just an `Element`/(+probably a `Widget`) subclass rather than implementing `Text`, `Radio`, etc?

I'm worried that subclassing `Element`  for every element/widget type that we have will end up with us writing a pretty massive amount of boilerplate while the tests themselves don't benefit much over more generic types. It could just look that way for now from the currently implemented tests, though.

AnOctopus replied:
At minimum there needs to be one for each widget, to translate interactions to the appropriate `WidgetState`, and provide higher level interaction methods. And there's enough individual variation with how to get values for other elements that probably most of them end up needing their own classes to work properly anyway.

AnOctopus committed:
Merge branch 'develop' into prototype/unit-test-interaction

AnOctopus committed:
Start moving interaction code into regular submodule

As preparation for putting test interface classes alongside the st
element function implementations.

AnOctopus committed:
Remove dead imports

AnOctopus committed:
Split out LocalScriptRunner

AnOctopus committed:
Fix some type complaints

AnOctopus committed:
Remove unused imports

AnOctopus committed:
More type fixes

AnOctopus committed:
Last pass of type error fixes

AnOctopus committed:
Remove unused imports

AnOctopus committed:
Add type alias for `Element | Block`

AnOctopus committed:
Add some interaction api docstrings

AnOctopus committed:
Add a little type information about widgets

Won't be able to get precise info because the widget type isn't
statically known, but at least editor autocomplete will know the method
exists.

AnOctopus committed:
Add comment about using Element directly

AnOctopus committed:
Get Protocol from typing_extensions for compatibility

AnOctopus committed:
Also import runtime_checkable from extensions

AnOctopus committed:
Merge branch 'develop' into prototype/unit-test-interaction

kajarenc commented on lib/streamlit/testing/element_tree.py:
```diff
@@ -0,0 +1,343 @@
+        ...
+
+    @overload
+    def get(self, elt: Literal["radio"]) -> Sequence[Radio[Any]]:
```
Could we have а little bit more extended name, something like `element_type` or `el_type` for the `elt` param name?

During the first read, I had the impression that `elt` is an abbreviation for `ELementT`. 
Feel free to ignore this comment if it is a common abbreviation

kajarenc commented:
looks good to me 👍 

There is still a lot of work to be done (I think in the list of remaining widgets and commands there will be many non-trivial corner cases that will have to be handled).

Another area that may need improvement is the script runner (how to run multi-page applications, running small scripts from a string seems not the most convenient to me, but I don’t have a better solution), as well as developing a full-fledged language for querying the widgets tree, but it is clear that it will be developed over time based on requests and feedback. 

I think now we are bound to file execution based on our current architecture and coupling our ScriptRunner with actual files on disk, maybe it would be great to have a better separation of concepts for file handling and streamlit code execution that could be part of this project or outside of the scope of this project. 

Thank you for your great work!

AnOctopus committed:
More explicit arg name for tree `get`
